src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs

```diff
@@ -5977,6 +5979,105 @@ public static Task<Task> WhenAny(params Task[] tasks)
             return TaskFactory.CommonCWAnyLogic(tasksCopy);
         }
 
+        // TODO https://github.com/dotnet/runtime/issues/23021: Make this public.
+        /// <summary>Creates a task that will complete when either of the supplied tasks have completed.</summary>
+        /// <param name="task1">The first task to wait on for completion.</param>
+        /// <param name="task2">The second task to wait on for completion.</param>
+        /// <returns>A task that represents the completion of one of the supplied tasks.  The return Task's Result is the task that completed.</returns>
+        /// <remarks>
+        /// The returned task will complete when any of the supplied tasks has completed.  The returned task will always end in the RanToCompletion state
+        /// with its Result set to the first task to complete.  This is true even if the first task to complete ended in the Canceled or Faulted state.
+        /// </remarks>
+        /// <exception cref="System.ArgumentNullException">
+        /// The <paramref name="task1"/> or <paramref name="task2"/> argument was null.
+        /// </exception>
+        internal static Task<Task> WhenAny(Task task1, Task task2) =>
+            (task1 is null) || (task2 is null) ? throw new ArgumentNullException(task1 is null ? nameof(task1) : nameof(task2)) :
+            task1.IsCompleted ? FromResult(task1) :
+            task2.IsCompleted ? FromResult(task2) :
+            new TwoTaskWhenAnyPromise<Task>(task1, task2);
+
+        /// <summary>A promise type used by WhenAny to wait on exactly two tasks.</summary>
+        /// <typeparam name="TTask">Specifies the type of the task.</typeparam>
+        /// <remarks>
+        /// This has essentially the same logic as <see cref="TaskFactory.CompleteOnInvokePromise"/>, but optimized
+        /// for two tasks rather than any number. Exactly two tasks has shown to be the most common use-case by far.
+        /// </remarks>
+        private sealed class TwoTaskWhenAnyPromise<TTask> : Task<TTask>, ITaskCompletionAction where TTask : Task
+        {
+            private TTask? _task1, _task2;
+
+            /// <summary>Instantiate the promise and register it with both tasks as a completion action.</summary>
+            public TwoTaskWhenAnyPromise(TTask task1, TTask task2)
+            {
+                Debug.Assert(task1 != null && task2 != null);
+                _task1 = task1;
+                _task2 = task2;
+
+                if (AsyncCausalityTracer.LoggingOn)
+                {
+                    AsyncCausalityTracer.TraceOperationCreation(this, "Task.WhenAny");
+                }
+
+                if (s_asyncDebuggingEnabled)
+                {
+                    AddToActiveTasks(this);
+                }
+
+                task1.AddCompletionAction(this);
+
+                task2.AddCompletionAction(this);
+                if (task1.IsCompleted)
```

Comment:
Is it worth it to check rask2.IsCompleted too?

Reply:
We could check task2, but it would actually be for the inverse condition.  We're checking task1 here because if it happened to complete before we registered the continuation with task2, we could be leaking the continuation into task2 (the code that ran when task1 completed would have tried to remove it from task2 but found it's not yet there).  But when task2 completes, it's going to clear its continuation list, so it handles any race condition with regards to adding the continuation.  If we wanted to add a check, it would simply be to avoid the extra call to task2.RemoveContinuation as an optimization, e.g. `if (task1.IsCompleted && !task2.IsCompleted)`, but we already did such a check prior to this code running, and RemoveContinuation does such a check, so there's only a tiny window where it would add benefit.  I'll leave it as is.